Patch 1

2Dman opened the pull request:
When using 'get-cypolicy' the object resturen policy_id instead of id.
To workaround this issue, we can add the id property:
$PolicyObject | Add-Member -Name id -Value $PolicyObject.policy_id -MemberType NoteProperty

- Or change the  Set-CyPolicyForDevice to expect a policy_id

2Dman committed:
Cypolicies.ps1

When using the module:
the get-cypolicy returns a policy_id property instead of an ID property

2Dman committed:
Cy-Policies.ps1

Changed Json object from id to policy_id, since get-cypolicy return a policy_id property and not an id property

jan-tee commented:
The objects are exactly the same structure as the API returns them - CyCLI is intentionally a very thin layer between API and Powershell, so that the API docs are applicable and the primitives work the same way as the functions in the API. Set-CyPolicyForDevice, like all other functions, expects a "policy" object. The problem here is caused by the policy being represented by the API in two different forms:

1) one where policy_id contains its ID (you discovered this one, but assumed this was the only form), returned by "get list of policies"
2) one where id contains its ID, returned by "get details of a policy".

Your patch would solve support for case 1 at the expense of case 2.

A "good" patch would need to support both forms. Do you want to revise this and I'll merge, or do you want me to address it with a patch that covers both cases?

2Dman committed:
Cypolicies.ps1

rollback: changed policy_id --> id
added logic to check if id or logic_id is present, and if it is: is not null, and/or the value are the same

It would be less complex if the get-cypolicy would fill up the id in the get- stage like: 
$PolicyObject | Add-Member -Name id -Value $PolicyObject.policy_id -MemberType NoteProperty

jan-tee commented:
Changing the policy object (by adding a property) is against the contract of keeping the objects as they come. I don't want to introduce objects that have synthetic properties, otherwise I would have done this in tens of places already where I disagree with the utility of the built-in objects (for startes, I would have added LDAP style "object class" attributes short of having strongly typed objects, and indicated shallow vs. full objects everywhere, etc.). IOW check first if an ID is there (either id or policy_id), store in method-scope variable, and then use that for the update call.

2Dman committed:
Update CyPolicies.ps1

changed to use variable in function for policy.policy_id or policy.id

jan-tee commented:
Looks good! Congratulations!
